### DISCUSSION TITLE
next-safe-action v6

### DISCUSSION THREAD

**Discussion by TheEdoRan** [next-safe-action/next-safe-action#47] [Announcements]
# next-safe-action v6

## What's new?

With next-safe-action version 6, you can now use a wide range of validation libraries, even multiple and custom ones at the same time, thanks to the great [TypeSchema](https://typeschema.com/) library. You can find supported libraries [here](https://typeschema.com/#coverage).

Existing code will not be affected, since Zod is supported by TypeSchema. However, now you can for example define a new safe action using [Yup](https://github.com/jquense/yup) or [Valibot](https://valibot.dev/), while still keeping existing actions with Zod validation, and everything will be handled internally by next-safe-action, thanks to the TypeSchema abstractions.

### Action result object

- Property `validationError` is now called `validationErrors`.

### Safe action client

- `handleReturnedServerError()` function now directly returns the server error message as a `string`, instead of a
`{ serverError: string }` object.

### Hooks

Hooks are now exported from `next-safe-action/hooks` instead of `next-safe-action/hook`.

### Types

- `ServerCode` is now called `ServerCodeFn`.

### Types

- Exported new `SafeClientOpts` type, which represents the options for the safe action client, used internally by `createSafeActionClient()` function.

**Comment by Myks92** [comment]
@TheEdoRan, what if we add another handler that is named as `handleReturnedServerValidationErrors'? Why would it be necessary? For example, when we send server actions to some third-party API resource, it can issue validation on the server side of the form:

```
{
    "errors": {
        "name": "The value type must be string.",
        "redirectUri": "The value type must be string.",
        "scope": "The value type must be string.",
        "grandType": "The value type must be string."
    }
}
```

In the current version, we cannot process this in any way, but we would like to... Tell me, is it possible? Can you do server validation error handling?

**Reply by TheEdoRan** [reply]
Is this solved by #52?

**Comment by naquiroz** [comment]
How can I use valibot in a form action @TheEdoRan ?
